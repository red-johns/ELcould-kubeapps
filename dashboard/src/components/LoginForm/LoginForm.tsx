import { Location } from "history";
import * as React from "react";
import { Lock } from "react-feather";
import { Redirect } from "react-router";

import "./LoginForm.css";

interface ILoginFormProps {
  authenticated: boolean;
  authenticating: boolean;
  authenticationError: string | undefined;
  authenticate: (token: string) => any;
  location: Location;
}

interface ILoginFormState {
  token: string;
}

class LoginForm extends React.Component<ILoginFormProps, ILoginFormState> {
  public state: ILoginFormState = { token: "" };
  public render() {
    if (this.props.authenticated) {
      const { from } = this.props.location.state || { from: { pathname: "/" } };
      return <Redirect to={from} />;
    }
    return (
      <section className="LoginForm">
        <div className="LoginForm__container padding-v-bigger bg-skew">
          <div className="container container-tiny">
            {this.props.authenticationError && (
              <div className="alert alert-error margin-c" role="alert">
                There was an error connecting to the Kubernetes API. Please check that your token is
                valid.
              </div>
            )}
          </div>
          <div className="bg-skew__pattern bg-skew__pattern-dark type-color-reverse">
            <div className="container">
              <h2>
                <Lock /> Login
              </h2>
              <p>
                Your cluster operator should provide you with a Kubernetes API token.{" "}
                <a
                  href="https://github.com/kubeapps/kubeapps/blob/master/docs/user/access-control.md"
                  target="_blank"
                >
                  Click here
                </a>{" "}
                for more info on how to create and use Bearer Tokens.
              </p>
              <div className="bg-skew__content">
                <form onSubmit={this.handleSubmit}>
                  <div>
                    <label htmlFor="token">Kubernetes API Token</label>
                    <input
                      name="token"
                      id="token"
                      type="password"
                      placeholder="Token"
                      required={true}
                      onChange={this.handleTokenChange}
                      value={this.state.token}
                    />
                  </div>
                  <p>
                    <button
                      type="submit"
                      className="button button-accent"
                      disabled={this.props.authenticating}
                    >
                      Login
                    </button>
                  </p>
                </form>
              </div>
            </div>
          </div>
        </div>
      </section>
    );
  }

  private handleSubmit = async (e: React.FormEvent<HTMLFormElement>) => {
    e.preventDefault();
    const { token } = this.state;
    return token && (await this.props.authenticate(token));
  };

  private handleTokenChange = (e: React.FormEvent<HTMLInputElement>) => {
    this.setState({ token: e.currentTarget.value });
  };
}

export default LoginForm;
