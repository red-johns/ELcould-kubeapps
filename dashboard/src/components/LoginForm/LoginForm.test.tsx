import { shallow } from "enzyme";
import { Location } from "history";
import * as React from "react";
import { Redirect } from "react-router-dom";

import LoginForm from "./LoginForm";

const emptyLocation: Location = {
  hash: "",
  pathname: "",
  search: "",
  state: "",
};

const defaultProps = {
  authenticate: jest.fn(),
  authenticated: false,
  authenticating: false,
  authenticationError: undefined,
  location: emptyLocation,
};

const authenticationError = "it's a trap";

it("renders a token login form", () => {
  const wrapper = shallow(<LoginForm {...defaultProps} />);
  expect(wrapper.find("input#token").exists()).toBe(true);
  expect(wrapper.find(Redirect).exists()).toBe(false);
  expect(wrapper).toMatchSnapshot();
});

it("renders a link to the access control documentation", () => {
  const wrapper = shallow(<LoginForm {...defaultProps} />);
  expect(wrapper.find("a").props()).toMatchObject({
    href: "https://github.com/kubeapps/kubeapps/blob/master/docs/user/access-control.md",
    target: "_blank",
  });
});

it("updates the token in the state when the input is changed", () => {
  const wrapper = shallow(<LoginForm {...defaultProps} />);
  wrapper.find("input#token").simulate("change", { currentTarget: { value: "f00b4r" } });
  expect(wrapper.state("token")).toBe("f00b4r");
});

describe("redirect if authenticated", () => {
  it("redirects to / if no current location", () => {
    const wrapper = shallow(<LoginForm {...defaultProps} authenticated={true} />);
    const redirect = wrapper.find(Redirect);
    expect(redirect.exists()).toBe(true);
    expect(redirect.props()).toEqual({ push: false, to: { pathname: "/" } });
  });

  it("redirects to previous location", () => {
    const location = Object.assign({}, emptyLocation);
    location.state = { from: "/test" };
    const wrapper = shallow(
      <LoginForm {...defaultProps} authenticated={true} location={location} />,
    );
    const redirect = wrapper.find(Redirect);
    expect(redirect.exists()).toBe(true);
    expect(redirect.props()).toEqual({ push: false, to: "/test" });
  });
});

it("calls the authenticate handler when the form is submitted", () => {
  const wrapper = shallow(<LoginForm {...defaultProps} />);
  wrapper.find("input#token").simulate("change", { currentTarget: { value: "f00b4r" } });
  wrapper.find("form").simulate("submit", { preventDefault: jest.fn() });
  expect(defaultProps.authenticate).toBeCalledWith("f00b4r");
});

it("displays an error if the authentication error is passed", () => {
  const wrapper = shallow(
    <LoginForm {...defaultProps} authenticationError={authenticationError} />,
  );

  expect(wrapper.find(".alert-error").exists()).toBe(true);
  expect(wrapper).toMatchSnapshot();
});

it("disables the input if authenticating", () => {
  const wrapper = shallow(<LoginForm {...defaultProps} />);

  expect(wrapper.find(".button").props().disabled).toBe(false);
  wrapper.setProps({ authenticating: true });
  expect(wrapper.find(".button")).toBeDisabled();
});
