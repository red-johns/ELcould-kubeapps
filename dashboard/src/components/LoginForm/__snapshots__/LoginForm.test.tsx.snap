// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`displays an error if the authentication error is passed 1`] = `
<section
  className="LoginForm"
>
  <div
    className="LoginForm__container padding-v-bigger bg-skew"
  >
    <div
      className="container container-tiny"
    >
      <div
        className="alert alert-error margin-c"
        role="alert"
      >
        There was an error connecting to the Kubernetes API. Please check that your token is valid.
      </div>
    </div>
    <div
      className="bg-skew__pattern bg-skew__pattern-dark type-color-reverse"
    >
      <div
        className="container"
      >
        <h2>
          <Lock
            color="currentColor"
            size="24"
          />
           Login
        </h2>
        <p>
          Your cluster operator should provide you with a Kubernetes API token.
           
          <a
            href="https://github.com/kubeapps/kubeapps/blob/master/docs/user/access-control.md"
            target="_blank"
          >
            Click here
          </a>
           
          for more info on how to create and use Bearer Tokens.
        </p>
        <div
          className="bg-skew__content"
        >
          <form
            onSubmit={[Function]}
          >
            <div>
              <label
                htmlFor="token"
              >
                Kubernetes API Token
              </label>
              <input
                id="token"
                name="token"
                onChange={[Function]}
                placeholder="Token"
                required={true}
                type="password"
                value=""
              />
            </div>
            <p>
              <button
                className="button button-accent"
                disabled={false}
                type="submit"
              >
                Login
              </button>
            </p>
          </form>
        </div>
      </div>
    </div>
  </div>
</section>
`;

exports[`renders a token login form 1`] = `
<section
  className="LoginForm"
>
  <div
    className="LoginForm__container padding-v-bigger bg-skew"
  >
    <div
      className="container container-tiny"
    />
    <div
      className="bg-skew__pattern bg-skew__pattern-dark type-color-reverse"
    >
      <div
        className="container"
      >
        <h2>
          <Lock
            color="currentColor"
            size="24"
          />
           Login
        </h2>
        <p>
          Your cluster operator should provide you with a Kubernetes API token.
           
          <a
            href="https://github.com/kubeapps/kubeapps/blob/master/docs/user/access-control.md"
            target="_blank"
          >
            Click here
          </a>
           
          for more info on how to create and use Bearer Tokens.
        </p>
        <div
          className="bg-skew__content"
        >
          <form
            onSubmit={[Function]}
          >
            <div>
              <label
                htmlFor="token"
              >
                Kubernetes API Token
              </label>
              <input
                id="token"
                name="token"
                onChange={[Function]}
                placeholder="Token"
                required={true}
                type="password"
                value=""
              />
            </div>
            <p>
              <button
                className="button button-accent"
                disabled={false}
                type="submit"
              >
                Login
              </button>
            </p>
          </form>
        </div>
      </div>
    </div>
  </div>
</section>
`;
